### ISSUE TITLE
`resume=true` continue training model with new data

### ISSUE THREAD

**Issue by derek-knox** [ultralytics/ultralytics#4042]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I read the [Resuming Interrupted Trainings](https://docs.ultralytics.com/modes/train/#resuming-interrupted-trainings) and I have a few questions regarding:
> This can come in handy in various scenarios, like when the training process has been unexpectedly interrupted, or when you wish to continue training a model with new data or for more epochs.

I've confirmed `resume=true` worked for the "training process has been unexpectedly interrupted" case, so thank you for this feature.

The documentation is great, but I'm a little unsure about a few things still. Below are my questions regarding `resume=true` "when you wish to continue training a model with new data":
1. What is the expected usage if the previous model had finished properly vs. unexpected interuption?
1. Do the  `train|val/labels.cache` files need to be deleted manually after adding new data to the dataset?
1. Does the same command for the original training need to now include both `resume=true` and `epoch=some-larger-number-than-the-original-epoch-count`?
1. If there is anything else that sounds pertinent given my other questions please add more context.

**Comment by aypetrunin** [comment]
Hi! I'll add my question. How to continue training the model if the previous training ended correctly? Relevant for free Golab, when training time is limited and forced to set a small number of epochs?
And why are the models saved after each epoch four times the size of the last and best model?

**Comment by glenn-jocher** [comment]
@derek-knox @aypetrunin if a `best.pt` model has completed successfully then you can use it as pretrained weights for a new training very easily:

```
yolo train model=path/to/best.pt ...
```

You never need to modify *.cache files, they are handled automatically.

**Comment by derek-knox** [comment]
@glenn-jocher

Using the `best.pt` as pretrained weights make total sense. I guess the "when you wish to continue training a model with new data" in combo with the `resume=true` doesn't make sense to me. It seems that portion of the docs should be removed in favor of simply using the `best.pt` as the pretrained model in a normal train run. Is that accurate, or is there a subtlety that I'm missing?

Thanks for confirming that the `*.cache` files never need to be touched manually.

**Comment by glenn-jocher** [comment]
@derek-knox In the context of `resume=true`, continuing training with new data means if you had an interrupted session but have added more data to your dataset, the model will continue training on the new combined dataset.

However, if you have a completely trained model, and you want to train it further on a new data, your understanding is correct. You would just call a normal train run specifying the `best.pt` from the previous training as your pretrained weights.

The `resume=true` flag and `best.pt` weights fulfil slightly different use-cases, hence both options are available.

**Comment by aypetrunin** [comment]
@glenn-jocher, If we continue to train the model on the same dataset using the **best.pt** model, are the elements of the dataset taken **randomly** or will they be **repeated**? Or change seed = 0 to seed = 1?

**Comment by glenn-jocher** [comment]
@aypetrunin hello,

If you continue training the model using the `best.pt` on the same dataset, the elements of the dataset are shuffled at the beginning of each epoch, so they are not repeated in the same order. 

As for the seed, it does not automatically change when you resume training. If you want to use a different seed you would have to set it manually in your training command. The seed helps in determining the randomization of data like the shuffle process of the dataset, and other random elements of the training.

I hope this answers your queries. Feel free to reach out if you have any more questions.

**Comment by aypetrunin** [comment]
@glenn-jocher, Hi. Thank you very much for your replies and generosity! How many questions I have! ))))

**Comment by derek-knox** [comment]
@glenn-jocher Thank you for confirming.

**Comment by glenn-jocher** [comment]
@derek-knox you're most welcome! I'm here to help answer your questions, and I'm glad I could assist. Don't hesitate to reach out if you have any more queries about YOLOv8. Your engagement helps us make our software better. Happy coding!

**Comment by derek-knox** [comment]
@glenn-jocher I have a follow-up to [your previous](https://github.com/ultralytics/ultralytics/issues/4042#issuecomment-1657668224):

> However, if you have a completely trained model, and you want to train it further on a new data, your understanding is correct. You would just call a normal train run specifying the `best.pt` from the previous training as your pretrained weights.

My questions are:
1. When deciding to train a model based off an existing YOLOn|s|etc. vs. a `best.pt` _when a dataset is expanded upon_, what are the reasons to take one approach over another?
    - Put another way, I know training off a `best.pt` would likely save time compared to running from scratch using YOLOn|s|etc. on an updated dataset, but what I don't intuit yet are the tradeoffs.
1. Are there any general guidelines informing when to train off a base model vs. expanding on a domain model?

**Comment by glenn-jocher** [comment]
@derek-knox great questions!

1. When a dataset is expanded, the choice of continuing training from `best.pt` versus restarting training from a pre-existing YOLO model depends on how closely your new data follows the same pattern as your original data. If the new data is very similar to the original, it could be beneficial to use the `best.pt` file and continue training from there, as the model is already well-optimized for the characteristics of your dataset. 

   However, if the new data is significantly different in characteristics (like different object categories, different lighting conditions, different backgrounds, etc), then it might lead to your model being overly biased towards the original data. This is essentially 'catastrophic forgetting,' where a neural network fails to maintain the information it has learned when new knowledge is incorporated. In that case, you might want to restart training from scratch using a pre-existing YOLO model.

2. Generally, it's beneficial to start from a model that's pre-trained on a large-scale and diverse dataset like ImageNet when you have limited data. This approach, also known as transfer learning, allows the model to use the learnt feature extractors from ImageNet as a generic starting point, which can be fine-tuned on your data.

   However, as your dataset becomes larger and more unique, it might become more advantageous to train from scratch or use a domain model. This model is optimized for the specific characteristics of your dataset and could potentially capture intricate nuances that a generic pre-trained model might miss. 

So, the decision is usually a trade-off between training time and performance. Training from `best.pt` would almost always be faster, but might not always yield the best performance, depending on the characteristics of your new data.

**Comment by derek-knox** [comment]
@glenn-jocher Thank you for the context, that is exactly what I was looking for.

**Comment by mansi-aggarwal-2504** [comment]
Hello, this is a helpful thread. But I want to make sure I understand it correctly for I want to do.
I have a fully trained model (75 epochs) on a certain dataset and I tested it on another dataset. I now want to further train my model on this test dataset. Should I just run:

`python train.py --data test.yaml --epochs 50 --weights 'path/to/best_75_epochs.pt'`

Can I run this on fewer epochs or should it be the same as `best.pt`? And will this new model be considered to be trained with 75 epochs + new training epochs or just new training epochs on new data?

Thanks!

**Comment by glenn-jocher** [comment]
@mansi-aggarwal-2504 hello! You're on the right track. To further train your model on the new dataset, you can indeed use the command you've mentioned. It's perfectly fine to run this for fewer epochs since you're continuing the training process. The new model will effectively have the knowledge from the initial 75 epochs plus whatever additional epochs you train it for on the new data. Just make sure your `test.yaml` is set up correctly with the new dataset paths. Happy training! 😊🚀

**Comment by Khanhlinh0410** [comment]
@glenn-jocher  it'll be really nice if you can answer this question for me. I have used yolov8n.pt (the original one) to train my dataset 1 at first then I saved to best.pt. I want to use best.pt and resume = True to keep training on a new dataset, but I met the error: best.pt training to 30 epochs is finished, nothing to resume. How can I fix it? Thank you again

**Comment by glenn-jocher** [comment]
It sounds like the training process is recognizing that the model has already completed the specified number of epochs. To continue training on a new dataset, you should start a new training session using `

**Comment by Leanse1** [comment]
What if I have trained the model for 100 epochs and received certain mAP values, now if I have planned to further increase the epochs using same dataset. should i use best.pt or opt for resume=True with last.pt?

**Comment by glenn-jocher** [comment]
Use a new training run from your best weights; `resume=True` is for interrupted runs and locks most args (you can’t reliably extend `epochs` there). Recommended:

```bash
yolo detect train data=your_data.yaml model=path/to/best.pt epochs=200
```

If you specifically want to continue from the last-epoch weights instead of the
